src/graph/transform/union_partition.cc

```diff
@@ -8,6 +8,202 @@ using namespace dgl::runtime;
 
 namespace dgl {
 
+HeteroGraphPtr DisjointUnionHeteroGraph2(
+    GraphPtr meta_graph, const std::vector<HeteroGraphPtr>& component_graphs) {
+  CHECK_GT(component_graphs.size(), 0) << "Input graph list is empty";
+  std::vector<HeteroGraphPtr> rel_graphs(meta_graph->NumEdges());
+  std::vector<int64_t> num_nodes_per_type(meta_graph->NumVertices(), 0);
+
+  // Loop over all canonical etypes
+  auto format = component_graphs[0]->GetRelationGraph(0)->GetFormatInUse();
+  auto restrict_format = component_graphs[0]->GetRelationGraph(0)->GetRestrictFormat();
+  for (dgl_type_t etype = 0; etype < meta_graph->NumEdges(); ++etype) {
+    auto pair = meta_graph->FindEdge(etype);
+    const dgl_type_t src_vtype = pair.first;
+    const dgl_type_t dst_vtype = pair.second;
+    uint64_t src_offset = 0, dst_offset = 0;
+    HeteroGraphPtr rgptr = nullptr;
+
+    // do some preprocess
+    for (size_t i = 0; i < component_graphs.size(); ++i) {
+      const auto& cg = component_graphs[i];
+      CHECK_EQ(format, cg->GetRelationGraph(etype)->GetFormatInUse()) << \
+        "All batched graph should have same sparse format";
+      CHECK_EQ(restrict_format, cg->GetRelationGraph(etype)->GetRestrictFormat()) << \
+        "All batched graph should have same restrict_format format";
```

Comment:
That might be too strict requirements for the users. Since we are promoting opaque sparse formats, it would be hard (or tedious) for users specify formats for `dgl.batch`. How about we just directly pick one format that is available?

Reply:
Another question is what is the restricted format for the result graph? I would prefer no restriction since that is the most common case. In other word, I feel the restricted format option should stick with one graph and should not propagate to other graphs via operations.